File: app-python/entitlements_app/services/append_audit_event_service.py
"""Append an immutable audit event for a subscription or entitlement decision."""

from decimal import Decimal
import re

from fastapi import HTTPException, status

from entitlements_app.requests.append_audit_event_request import AppendAuditEventRequest
from entitlements_app.entities import AuditEvent
from entitlements_app.ports import EntitlementsRepository


class AppendAuditEventService:
    """Service handling Command `AppendAuditEvent`."""

    def __init__(self, entitlements_repository: EntitlementsRepository) -> None:
        self.entitlements_repository: EntitlementsRepository = entitlements_repository

    def handle(self, request: AppendAuditEventRequest) -> AuditEvent:
        # spec-driven body for Command AppendAuditEvent.
        # intent: Append an immutable audit event for a subscription or entitlement decision.
        # requires: EntitlementsRepository
        # return: AuditEvent (entity); construct from request inputs and persist via EntitlementsRepository.saveAuditEvent.
        value = AuditEvent(tenantId=request.tenantId, auditEventId=request.auditEventId, accountId=request.accountId, actorId=request.actorId, eventType=request.eventType, subjectId=request.subjectId, message=request.message, occurredAt=request.occurredAt)
        return self.entitlements_repository.saveAuditEvent(value)

File: app/src/main/java/com/example/spring/AppendAuditEventService.java
package com.example.spring;

import com.example.spring.error.ConflictException;
import com.example.spring.error.NotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Append an immutable audit event for a subscription or entitlement decision.
 */
@Service
public class AppendAuditEventService {

    private final EntitlementsRepository entitlementsRepository;

    public AppendAuditEventService(EntitlementsRepository entitlementsRepository) {
        this.entitlementsRepository = entitlementsRepository;
    }

    @Transactional
    public AuditEvent handle(AppendAuditEventRequest request) {
        // spec-driven body for AppendAuditEvent.
        // intent: Append an immutable audit event for a subscription or entitlement decision.
        // requires: EntitlementsRepository
        // return: AuditEvent (entity); spec-driven body constructs from request inputs + persists via EntitlementsRepository.saveAuditEvent.
        AuditEvent value = new AuditEvent(request.tenantId(), request.auditEventId(), request.accountId(), request.actorId(), request.eventType(), request.subjectId(), request.message(), request.occurredAt());
        return this.entitlementsRepository.saveAuditEvent(value);
    }
}